### ISSUE TITLE
Feature Request: Implement onBeforeCreate() and onAfterCreate()

### ISSUE THREAD

**Issue by madmatt** [silverstripe/silverstripe-framework#8258]
## Description

We have `onBeforeWrite()` and `onAfterWrite()` which is triggered on every write of an object to the database. We have `onBeforeDelete()` and `onAfterDelete()` which is triggered when an object is deleted from the database.

It'd be great to have an opposite on `onBeforeDelete()` and `onAfterDelete()` for when a `DataObject` is created for the first time. Enter:

### `onBeforeCreate()` and `onAfterCreate()`

These methods would be called when a `DataObject` is created, but *only the very first time it is created*. In other words, when you call `new Page()` or `Page::create()`. These extension points would enable modules to more easily affect core data classes. One key reason I can see for this is to set default properties that can't be set via `private static $defaults` - e.g. setting a content owner on a page to the current user.

**Comment by dhensby** [comment]
Hmm, on before create would be possible but on after seems hard? The only place this could go would be the constructor but any parent class that calls `parent::__construct()` would either have to implement their logic before the `beforeCreate` call or after the `afterCreate` call. So we'd have to tell people to stop using the `__construct()` method at all and use our internal methods.

The alternative would be for injector to literally do this instead (so not available with `new Page()` calls and just `Page::create()`) that way `beforeCreate` could literally do things *before* the object was created and after could do it literally *after*.

Other points I'd make:

1. I think the name should not be before/afterCreate as this would be much better as a hook for before or after database input for the first time
2. Your problem of updating values by default wouldn't be a problem, just override the `populateDefaults` method.

**Comment by maxime-rainville** [comment]
I'm with @dhensby. Putting this on the constructor feels a bit weird.

You sure you don't mean to have _on before/after the first write_? From the example you're describing that would make more sense.

**Comment by madmatt** [comment]
Well hey, I even managed to contradict myself in my very own feature request hehe.

There are two different things here:

1. Immediately before and after creating an object
2. Immediately before and after writing an object to the database for the first time

What I actually meant to suggest was option 2 - so the method could be more accurately described as `onBeforeFirstWrite()` and `onAfterFirstWrite()` for example.

However, I had forgotten about `populateDefaults()`, and that sort-of solves the use-case that I have. One other thing I was considering is for auditing purposes (knowing the difference between a new record being created for the first time and one being updated, without having to resort to fooling around with the `getChangedFields()` method and other random ways of figuring out if it's been written before or not.

**Comment by kinglozzer** [comment]
+1 for adding an extra pair of methods/extension points for first time writes that people can use rather than having to use `if ($this->isChanged('ID'))` or similar

**Comment by dhensby** [comment]
Yep, an `on(Before|After)Create` where "create" is first insertion to the DB would be helpful, but it isn't currently impossible at the moment as you *can* use `isChanged('ID')` as @kinglozzer points out

**Comment by maxime-rainville** [comment]
I think it would be a useful convenience method and it would have a nice symmetry with the other *onBefore/onAfter* methods.

**Comment by robbieaverill** [comment]
+1 for adding onBefore/AfterCreate as first write to DB - this is commonly used and would be a useful hook to have. I don't see much benefit in adding access to instance construction and destruction over what those methods would give you already.

**Comment by andrewandante** [comment]
As an aside, let's not call it `onBeforeCreate()` if it is actually `onBeforeFirstWrite()` - that's likely going to cause confusion and bite us down the line if we ever want to rename `populateDefaults()` to match the naming conventions, or add a separate `onBeforeCreate()` method.

> "I called `MyObject::create()` and the `onBeforeCreate()` hook didn't fire, what gives??"

~ Me, probably

**Comment by dhensby** [comment]
Yeah, I think @andrewandante is probably right here...

**Comment by NikxDa** [comment]
@maxime-rainville Tagging you as you were the person to accept this RFC. Is this still wanted? I might be able to pick this up sometime next month. :)

**Comment by maxime-rainville** [comment]
@NikxDa If you want to have a crack at implementing this, go for it. The idea has been accepted in-principle, so it's just a question of making sure we have a good implementation for it.

**Comment by meysamm22** [comment]
I think using `$this->isInDB()` condition in `onBeforeUpdate` event is a better idea.
